UBCSailbot/local-pathfinding
# GridStateSampler

This PR is still in progress. This PR builds from #189 . That was a good start, but for some reason, it was only making about 4-6 samples with 1 second... (maybe valid state sampler has some other overhead that regular state sampler doesn't have).

This PR adds:
* `num_runs`, `runtime_seconds`, `state_sampler`, `grid_n` parameters for easy testing and tuning
* Adds GridStateSampler that samples in a fixed x,y grid that is n x n (actually technically n+1 x n+1 with current implementation)

We want total pathfinding runtime to be about 1-2 seconds (higher might be too slow with upwind/downwind slowing us down)

To test:
`roslaunch local_pathfinding all.launch num_ais_ships:=25 random_seed:=310 runtime_seconds:=1.0 num_runs:=4 planner_type:=rrtstar state_sampler:="grid" grid_n:=10`

`roslaunch local_pathfinding all.launch num_ais_ships:=25 random_seed:=310 runtime_seconds:=1.0 num_runs:=4 planner_type:=rrtstar state_sampler:=""`

`roslaunch local_pathfinding all.launch num_ais_ships:=25 random_seed:=310 runtime_seconds:=2.0 num_runs:=2 planner_type:=prmstar state_sampler:="grid" grid_n:=15`

`roslaunch local_pathfinding all.launch num_ais_ships:=25 random_seed:=310 runtime_seconds:=4.0 num_runs:=1 planner_type:=rrtstar state_sampler:="grid" grid_n:=6`

Tune values for `grid_n` (int grid size), `state_sampler` ("grid" or "" for default), `planner_type` (rrtstar, etc. see allocatePlanner for more), `runtime_seconds` float time, `num_runs` number of planning runs (want num_runs * runtime_seconds <= 2 if possible), change number of ships and random seed to test different scenarios and difficulty levels.

Can look at `main_loop.py` `localPath.getCost()` or `rostopic echo /localPathCost` or just look at  the visualizer to evaluate quality.
